### PR TITLE
Use FedoraResource instead of Nodes (DEV branch)

### PR DESCRIPTION
**Pull request by whikloj**
https://jira.duraspace.org/browse/FCREPO-1965

Continuation of https://github.com/fcrepo4-exts/fcrepo-transform/pull/11

### CODE REVIEW

File: src/main/java/org/fcrepo/transform/http/FedoraTransform.java [hunk 1]
```diff
@@ -108,26 +105,20 @@ public FedoraTransform(final String externalPath) {
...
-                final Node baseConfig = node.addNode(NT_BASE, NT_FILE);
-                jcrTools.uploadFile(internalSession, baseConfig.getPath(), getClass().getResourceAsStream(
-                        "/ldpath/default/nt_base_ldpath_program.txt"));
+            // Create this resource or it becomes a PairTree which is not referenceable.
+            containerService.findOrCreate(internalSession, "/fedora:system/fedora:transform");
+            final FedoraResource resource =
+                    containerService.findOrCreate(internalSession, CONFIGURATION_FOLDER + "default");
+            LOGGER.debug("Transformation default resource: {}", resource.getPath());
+
+            if (!resource.getChildren().anyMatch(child -> child.getPath() == DEFAULT_TRANSFORM_RESOURCE)) {
```
Comment (ajs6f) [inline src/main/java/org/fcrepo/transform/http/FedoraTransform.java]: Aren't these `String`s? Are you really sure it's `==`?

File: src/main/java/org/fcrepo/transform/transformations/LDPathTransform.java [hunk 2]
```diff
@@ -78,60 +76,57 @@ public LDPathTransform(final InputStream query) {
...
-            if (programNode.hasNode(nodeType.toString())) {
-                return new LDPathTransform(programNode.getNode(nodeType.toString())
-                                               .getNode(JCR_CONTENT)
-                                               .getProperty(JCR_DATA)
-                                               .getBinary().getStream());
+        // convert rdf:type with URI namespace to prefixed namespace
+        final Function<URI, String> namespaceUriToPrefix = x -> {
+            final String uriString = x.toString();
+            try {
+                for (final String namespace : Arrays.asList(nsRegistry.getURIs())) {
```
Comment (ajs6f) [inline src/main/java/org/fcrepo/transform/transformations/LDPathTransform.java]: Why `Arrays.asList` with `for`? You can use `for Thing thing : things` with an array.

File: src/main/java/org/fcrepo/transform/transformations/LDPathTransform.java [hunk 3]
```diff
@@ -78,60 +76,57 @@ public LDPathTransform(final InputStream query) {
...
+                    // There is an empty namespace URI and that matches everything.
```
Comment (ajs6f) [inline src/main/java/org/fcrepo/transform/transformations/LDPathTransform.java]: I don't understand this comment. Given the test below, an empty ns clearly does _not_ match.

Comment (whikloj) [inline src/main/java/org/fcrepo/transform/transformations/LDPathTransform.java]: Perhaps my test is bad, but for some reason the namespace registry has a blank prefix as the first entry. That was matching. I'll take another look at it.

File: src/main/java/org/fcrepo/transform/transformations/LDPathTransform.java (hunk 2 above)
Comment (whikloj) [inline src/main/java/org/fcrepo/transform/transformations/LDPathTransform.java]: Good point, I'm sure I had a reason. Just not necessarily a good one.

File: src/main/java/org/fcrepo/transform/http/FedoraTransform.java (hunk 1 above)
Comment (whikloj) [inline src/main/java/org/fcrepo/transform/http/FedoraTransform.java]: Fair point, I'll fix that.

File: src/main/java/org/fcrepo/transform/transformations/LDPathTransform.java (hunk 3 above)
Comment (whikloj) [inline src/main/java/org/fcrepo/transform/transformations/LDPathTransform.java]: Ok, I played with this and the blank namespace does match.
With the `namespace.length() > 0` left in and some debug statements outside and inside the loop I see:

```
DEBUG 08:16:29.474 (LDPathTransform) namespace is 
DEBUG 08:16:29.474 (LDPathTransform) namespace is http://www.w3.org/1999/02/22-rdf-syntax-ns#
DEBUG 08:16:29.474 (LDPathTransform) namespace is http://www.jcp.org/jcr/mix/1.0
DEBUG 08:16:29.474 (LDPathTransform) namespace is http://www.jcp.org/jcr/1.0
DEBUG 08:16:29.474 (LDPathTransform) namespace is http://www.w3.org/ns/ldp#
DEBUG 08:16:29.474 (LDPathTransform) namespace is http://www.w3.org/2001/XMLSchema
DEBUG 08:16:29.474 (LDPathTransform) namespace is http://www.modeshape.org/images/1.0
DEBUG 08:16:29.475 (LDPathTransform) namespace is http://www.w3.org/2000/01/rdf-schema#
DEBUG 08:16:29.475 (LDPathTransform) namespace is http://purl.org/dc/elements/1.1/
DEBUG 08:16:29.475 (LDPathTransform) namespace is http://www.w3.org/2000/xmlns/
DEBUG 08:16:29.475 (LDPathTransform) namespace is http://www.ebu.ch/metadata/ontologies/ebucore/ebucore#
DEBUG 08:16:29.475 (LDPathTransform) namespace is http://www.loc.gov/premis/rdf/v1#
DEBUG 08:16:29.475 (LDPathTransform) namespace is http://www.jcp.org/jcr/sv/1.0
DEBUG 08:16:29.475 (LDPathTransform) namespace is info:fedora/test/
DEBUG 08:16:29.475 (LDPathTransform) namespace is http://fedora.info/definitions/v4/repository#
DEBUG 08:16:29.475 (LDPathTransform) namespace match on http://fedora.info/definitions/v4/repository# to URI http://fedora.info/definitions/v4/repository#Container
```

If I remove the `namespace.length() > 0` I end up with:

```
DEBUG 08:17:28.101 (LDPathTransform) namespace is ()
DEBUG 08:17:28.101 (LDPathTransform) namespace match on () to URI (http://fedora.info/definitions/v4/repository#Container)
```

Perhaps I shouldn't be using `.startsWith()`?

File: src/main/java/org/fcrepo/transform/http/FedoraTransform.java [hunk 4]
```diff
@@ -178,12 +175,12 @@ public Object evaluateLdpathProgram(@PathParam("program") final String program)
     @POST
     @Consumes({APPLICATION_RDF_LDPATH, contentTypeSPARQLQuery})
     @Produces({APPLICATION_JSON, contentTypeTextTSV, contentTypeTextCSV,
-            contentTypeSSE, contentTypeTextPlain, contentTypeResultsJSON,
-            contentTypeResultsXML, contentTypeResultsBIO, contentTypeTurtle,
-            contentTypeN3, contentTypeNTriples, contentTypeRDFXML})
+        contentTypeSSE, contentTypeTextPlain, contentTypeResultsJSON,
```
Comment (awoods) [inline src/main/java/org/fcrepo/transform/http/FedoraTransform.java]: Please rollback indentation updates.

File: src/main/java/org/fcrepo/transform/http/FedoraTransform.java [hunk 5]
```diff
@@ -178,12 +175,12 @@ public Object evaluateLdpathProgram(@PathParam("program") final String program)
...
+        contentTypeResultsXML, contentTypeResultsBIO, contentTypeTurtle,
+        contentTypeN3, contentTypeNTriples, contentTypeRDFXML})
     @Timed
     public Object evaluateTransform(@HeaderParam("Content-Type") final MediaType contentType,
-                                    final InputStream requestBodyStream) {
```
Comment (awoods) [inline src/main/java/org/fcrepo/transform/http/FedoraTransform.java:188]: Please rollback indentation updates.

File: src/main/java/org/fcrepo/transform/transformations/LDPathTransform.java [hunk 6]
```diff
@@ -78,60 +74,57 @@ public LDPathTransform(final InputStream query) {
...
+                .findFirst()
+                .orElseThrow(() -> new TransformNotFoundException(String.format("Couldn't find transformation for {} and transformation key {}", resource.getPath(), key)));
+        return new LDPathTransform(transform.getContent());
     }
 
     @Override
     public List<Map<String, Collection<Object>>> apply(final RdfStream stream) {
         final LDPath<RDFNode> ldpathForResource =
-            getLdpathResource(stream);
+                getLdpathResource(stream);
```
Comment (awoods) [inline src/main/java/org/fcrepo/transform/transformations/LDPathTransform.java]: Please rollback indentation updates.

File: src/main/java/org/fcrepo/transform/transformations/LDPathTransform.java [hunk 7]
```diff
@@ -78,60 +74,57 @@ public LDPathTransform(final InputStream query) {
...
 
         final Resource context = createResource(stream.topic().getURI());
 
         try {
             return ImmutableList.of(unsafeCast(
-                ldpathForResource.programQuery(context, new InputStreamReader(query))));
+                    ldpathForResource.programQuery(context, new InputStreamReader(query))));
```
Comment (awoods) [inline src/main/java/org/fcrepo/transform/transformations/LDPathTransform.java]: Please rollback indentation updates.

File: src/main/java/org/fcrepo/transform/transformations/LDPathTransform.java [hunk 8]
```diff
@@ -78,60 +74,57 @@ public LDPathTransform(final InputStream query) {
...
-     * @param node the node
-     * @param key the key
-     * @return node-type specific transform
-     * @throws RepositoryException if repository exception occurred
-     */
-    public static LDPathTransform getNodeTypeTransform(final Node node,
-        final String key) throws RepositoryException {
-
-        final Node programNode = node.getSession().getNode(CONFIGURATION_FOLDER + key);
+    public static LDPathTransform getResourceTransform(final FedoraResource resource, final NodeService nodeService,
```
Comment (awoods) [inline src/main/java/org/fcrepo/transform/transformations/LDPathTransform.java:85]: During the extraction of fcrepo-transform from the core codebase the checkstyle plugin was dropped. Please add a javadoc to this method in advance of that plugin being reintroduced.

File: src/main/java/org/fcrepo/transform/transformations/LDPathTransform.java [hunk 9]
```diff
@@ -78,60 +74,57 @@ public LDPathTransform(final InputStream query) {
...
+                .map(stringType -> transformResource.getPath() + "/" + stringType + "/jcr:content")
+                .collect(Collectors.toList());
 
-        throw new WebApplicationException(new Exception(
-                "Couldn't find transformation for " + node.getPath()
-                        + " and transformation key " + key), SC_BAD_REQUEST);
+        final FedoraBinary transform = (FedoraBinary) transformResource.getChildren()
+                .filter(child -> rdfStringTypes.contains(child.getPath()))
+                .findFirst()
+                .orElseThrow(() -> new TransformNotFoundException(String.format("Couldn't find transformation for {} and transformation key {}", resource.getPath(), key)));
```
Comment (awoods) [inline src/main/java/org/fcrepo/transform/transformations/LDPathTransform.java]: During the extraction of fcrepo-transform from the core codebase the checkstyle plugin was dropped. Please trim this line to 120 characters or less in advance of that plugin being reintroduced.

File: src/main/java/org/fcrepo/transform/TransformNotFoundException.java [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+    public TransformNotFoundException(final String msg) {
+        super(msg);
+    }
+
+    /**
+     * Ordinary constructor.
+     *
+     * @param rootCause the root cause
+     */
+    public TransformNotFoundException(final Throwable rootCause) {
```
Comment (awoods) [inline src/main/java/org/fcrepo/transform/TransformNotFoundException.java]: This constructor is never used. Please remove.

File: src/main/java/org/fcrepo/transform/TransformNotFoundException.java [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+        super(rootCause);
+    }
+
+    /**
+     * Ordinary constructor.
+     *
+     * @param msg the message
+     * @param rootCause the root cause
+     */
+    public TransformNotFoundException(final String msg, final Throwable rootCause) {
```
Comment (awoods) [inline src/main/java/org/fcrepo/transform/TransformNotFoundException.java]: This constructor is never used. Please remove.

**Comment by acoburn** [comment]
@whikloj can you rebase this PR against the `DEV` branch? That will help catch any checkstyle errors.

**Comment by whikloj** [comment]
@acoburn done, thanks for getting the checkstyle rules back in. 👍

**Pull request merged**